src/alpaka/AlpakaCore/alpakaWorkDivHelper.h

```diff
@@ -320,6 +320,495 @@ namespace cms {
       return isNextStrideElementValid;
     }
 
+    /*
+     * Class which simplifies "for" loops over elements index 
+     */
+    template <typename T, typename T_Acc>
+    class elements_with_stride {
+    public:
+
+    ALPAKA_FN_ACC elements_with_stride(const T_Acc& acc) {
+            const unsigned int dimIndex = 0;
+            const Idx threadIdxLocal(alpaka::getIdx<alpaka::Block, alpaka::Threads>(acc)[dimIndex]);
+            const Idx blockIdxInGrid(alpaka::getIdx<alpaka::Grid, alpaka::Blocks>(acc)[dimIndex]);
+
+            const Idx blockDimension(alpaka::getWorkDiv<alpaka::Block, alpaka::Elems>(acc)[dimIndex]);
+            const Idx gridDimension(alpaka::getWorkDiv<alpaka::Grid, alpaka::Elems>(acc)[dimIndex]);
+
+            thread_ = blockDimension * blockIdxInGrid + threadIdxLocal;
+            stride_ = blockDimension * gridDimension;
+            extent_ = stride_;
+            blockDim = blockDimension;
+        }
+
+    ALPAKA_FN_ACC elements_with_stride(const T_Acc& acc, T extent, const unsigned int dimIndex = 0) : extent_(extent) {
+            
+            const Idx threadIdxLocal(alpaka::getIdx<alpaka::Block, alpaka::Threads>(acc)[dimIndex]);
+            const Idx blockIdxInGrid(alpaka::getIdx<alpaka::Grid, alpaka::Blocks>(acc)[dimIndex]);
+
+            const Idx blockDimension(alpaka::getWorkDiv<alpaka::Block, alpaka::Elems>(acc)[dimIndex]);
+            const Idx gridDimension(alpaka::getWorkDiv<alpaka::Grid, alpaka::Elems>(acc)[dimIndex]);
+
+            thread_ = blockDimension * blockIdxInGrid + threadIdxLocal;
+            stride_ = blockDimension * gridDimension;
+            blockDim = blockDimension;
+        }
+    
+
+    class iterator {
+        friend class elements_with_stride;
+
+    public:
+        ALPAKA_FN_ACC constexpr T operator*() const { return index_; }
+
+        ALPAKA_FN_ACC constexpr iterator& operator++() {
+            
+            #ifdef ALPAKA_ACC_GPU_CUDA_ENABLED
+                // increment the index
+                index_ += stride_;
+                if (index_ < extent_)
+                    return *this;
+   
+            #else // CPU Backend
+                // Iterate over all the elements for one thread
+                index_ += 1;
+                if (index_ < old_index_ + blockDim) {
+                    return *this;
+                }
+            #endif
```

Comment:
(sorry this became long)

If I understood correctly, the point is that the kernel launch parameters would be expressed only in terms of number of blocks and number of threads (but not by number of elements), and then these are used differently for GPU and for CPU:
* For CUDA Alpaka's grid/block/thread are mapped to GPU grid/block/thread. The index is increased by the stride (= `blockDimension * gridDimension`), until the index hits the `extent_`
  * this leads to the "coalesced memory access"
* For Serial and TBB Alpaka's block is mapped to a CPU thread (on serial the whole grid is executed by one CPU thread, on TBB each block gets its own task (?)). The index is increased by one until the "end of the block" is reached.

In both cases the iterator traverses all indices of the range within all the CPU/GPU threads of the grid.

Omitting Alpaka's element level may become confusing if we'd end up having Alpaka code that uses the Alpaka indices directly, especially towards setting the kernel launch parameters. On the other hand, the goal here is to hide all the index calculations and "non-portability" of the launch parameters. But how sure can we be that such use cases either don't exist or are rare-enough that this disparity wouldn't matter?

I think there is sort-of a strategical question here (for which I don't have a clear opinion right now):
* Use the same (blocks, threads) launch parameters for all backends, and in the loop structure map the "threads" to threads on GPU and "blocks" on threads on CPU (as here), or
* Use (blocks, threads, elements) launch parameters, use the same loop structure for all backends, but provide helpers to map "abstract launch parameters" of (blocks, threads) to (blocks, threads, 1) for GPU and (blocks, 1, threads) on CPU (as in `alpakaWorkDivHelper.h`)

If we'd provide our own abstraction for the kernel launch, all this could be hidden from the user (as long as they wouldn't use `alpaka::getIdx()`). But I'm not suggesting to do that.

One question that could be tested comes to my mind. To my understanding one of the main points of the "loop over elements" is to auto-vectorize the loop on CPU. It could be interesting to test how all these abstraction attempts affect the auto-vectorization capabiliies of the compiler (GCC might be enough, also clang could be interesting).

Reply:
Indeed, this is what I have done.

> For Serial and TBB Alpaka's block is mapped to a CPU thread (on serial the whole grid is executed by one CPU thread, on TBB each block gets its own task (?)). The index is increased by one until the "end of the block" is reached.

I think that I must add here a condition which checks if the index is greater than ```extent_``` as well, not only the one with the "end of the block".

> To my understanding one of the main points of the "loop over elements" is to auto-vectorize the loop on CPU. It could be interesting to test how all these abstraction attempts affect the auto-vectorization capabiliies of the compiler (GCC might be enough, also clang could be interesting).

Yes, we can discuss more about this after I will make all the changes to the classes.